auth/jwt/middleware.go

```diff
@@ -0,0 +1,107 @@
+package jwt
+
+import (
+	"errors"
+
+	"golang.org/x/net/context"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/go-kit/kit/endpoint"
+)
+
+const (
+	// JWTTokenContextKey holds the key used to store a JWT Token in the context
+	JWTTokenContextKey = "JWTToken"
+	// JWTClaimsContxtKey holds the key used to store the JWT Claims in the context
+	JWTClaimsContextKey = "JWTClaims"
+)
+
+var (
+	ErrTokenContextMissing     = errors.New("Token up for parsing was not passed through the context")
+	ErrTokenInvalid            = errors.New("JWT Token was invalid")
+	ErrUnexpectedSigningMethod = errors.New("Unexptected signing method")
+	ErrKIDNotFound             = errors.New("Key ID was not found in key set")
+	ErrNoKIDHeader             = errors.New("Token doesn't have 'kid' header")
+)
+
+type Claims map[string]interface{}
+
+type KeySet map[string]struct {
+	Method jwt.SigningMethod
+	Key    []byte
+}
+
+// Create a new JWT token generating middleware, specifying signing method and the claims
+// you would like it to contain. Particularly useful for clients.
+func NewSigner(kid string, keys KeySet, claims Claims) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			key, ok := keys[kid]
+			if !ok {
+				return nil, ErrKIDNotFound
+			}
+
+			token := jwt.NewWithClaims(key.Method, jwt.MapClaims(claims))
+			token.Header["kid"] = kid
+			// Sign and get the complete encoded token as a string using the secret
+			tokenString, err := token.SignedString(key.Key)
+			if err != nil {
+				return nil, err
+			}
+			ctx = context.WithValue(ctx, JWTTokenContextKey, tokenString)
+
+			return next(ctx, request)
+		}
+	}
+}
+
+// Create a new JWT token parsing middleware, specifying a jwt.Keyfunc interface and the
+// signing method. Adds the resulting claims to endpoint context or returns error on invalid
+// token. Particularly useful for servers.
+func NewParser(keys KeySet) endpoint.Middleware {
```

Comment:
Hey @cam-stitt, what is included in your keyfunc? I was trying to build an interface that allowed for easy use of the `KID` (Key ID) header as means of identifying which key to use to verify a token. If it makes more sense to have added flexibility via a keyfunc then I can revert back to using that. (Comment hasn't been updated since i made this last change) 

See code at https://github.com/go-kit/kit/blob/1961299ceed1e99ddb9a868a51cd2177cf98583b/auth/jwt/middleware.go to see the keyfunc implementation.